pkg/plugins/gateway/algorithms/vtc/token_tracker.go

```diff
@@ -119,79 +155,224 @@ func (t *InMemorySlidingWindowTokenTracker) getCutoffTimestamp() int64 {
+	if user == "" {
+		return fmt.Errorf("user ID cannot be empty")
+	}
+
 	t.mu.Lock()
 	defer t.mu.Unlock()
 
-	if user == "" {
-		return nil
-	}
 	now := time.Now()
+	currentTimestamp := t.bucketUnit.toTimestamp(now)
+	cutoff := t.getCutoffTimestamp()
 
-	truncatedTime := t.bucketUnit.truncateTime(now)
-	windowStart := t.bucketUnit.toTimestamp(truncatedTime)
-
-	tokens := inputTokens*t.config.InputTokenWeight + outputTokens*t.config.OutputTokenWeight
-	buckets, ok := t.userBuckets[user]
+	_, ok := t.userBucketStore[user]
 	if !ok {
-		buckets = make(map[int64]float64)
-		t.userBuckets[user] = buckets
+		t.userBucketStore[user] = &userBucketData{
+			buckets: list.New(),
+			lookup:  make(map[int64]*list.Element),
+		}
 	}
-	buckets[windowStart] += tokens
 
-	cutoff := t.getCutoffTimestamp()
-	t.pruneExpiredBuckets(user, cutoff)
+	oldTotal := t.userTotals[user]
+
+	// Prune first before adding/updating to maintain window size constraint accurately
+	t.pruneExpiredBucketsAndUpdateState(user, cutoff)
+
+	totalAfterPruning := t.userTotals[user]
+	newTokens := inputTokens*t.config.InputTokenWeight + outputTokens*t.config.OutputTokenWeight
+
+	// Check if a bucket for the current timestamp already exists
+	if element, exists := t.userBucketStore[user].lookup[currentTimestamp]; exists {
+		node := element.Value.(*bucketNode)
+		node.tokenCount += newTokens
+	} else {
+		node := &bucketNode{timestamp: currentTimestamp, tokenCount: newTokens}
+		element := t.userBucketStore[user].buckets.PushBack(node)
+		t.userBucketStore[user].lookup[currentTimestamp] = element
+	}
+
+	// Calculate the final new total
+	newTotal := totalAfterPruning + newTokens
+
+	// Update user total and global min/max efficiently
+	t.updateUserTotalAndMinMax(user, oldTotal, newTotal)
 
 	return nil
 }
+
+func (t *InMemorySlidingWindowTokenTracker) updateUserTotalAndMinMax(user string, oldTotal, newTotal float64) {
+	if oldTotal == newTotal {
+		return
+	}
+	if newTotal < 0 {
+		newTotal = 0
+	}
+	t.userTotals[user] = newTotal
+	t.removeFromTotals(oldTotal, user)
+	t.addToTotals(newTotal, user)
+	t.recalcMin(oldTotal, newTotal)
+	t.recalcMax(oldTotal, newTotal)
+}
+
+func (t *InMemorySlidingWindowTokenTracker) removeFromTotals(oldTotal float64, user string) {
+	if oldTotal <= 0 {
+		return
+	}
+	if users, ok := t.totalsToUsers[oldTotal]; ok {
+		delete(users, user)
+		if len(users) == 0 {
+			delete(t.totalsToUsers, oldTotal)
+		}
+	}
+}
+
+func (t *InMemorySlidingWindowTokenTracker) addToTotals(newTotal float64, user string) {
+	if newTotal <= 0 {
+		return
+	}
+	if _, ok := t.totalsToUsers[newTotal]; !ok {
+		t.totalsToUsers[newTotal] = make(map[string]struct{})
+	}
+	t.totalsToUsers[newTotal][user] = struct{}{}
+}
+
+func (t *InMemorySlidingWindowTokenTracker) recalcMin(oldTotal, newTotal float64) {
+	if oldTotal > 0 && oldTotal == t.minTrackedToken && len(t.totalsToUsers[oldTotal]) == 0 {
+		t.minTrackedToken = math.Inf(1)
```

Comment:
Can we simplify the logic a little bit here? is `t.minTrackedToken` used somewhere else? 

```
t.minTrackedToken = math.MaxFloat64
for total := range t.totalsToUsers {
    if total < t.minTrackedToken {
        t.minTrackedToken = total
    }
}
```

Reply:
Yes, good suggestion